src/listen/heosResponse.ts

```diff
@@ -0,0 +1,40 @@
+import { HeosResponse, HeosResponseMessage } from '../types'
+import {
+	FIRST_LEVEL_HEOS_RESPONSE_MESSAGE_SPLITTER,
+	SECOND_LEVEL_HEOS_RESPONSE_MESSAGE_SPLITTER
+} from '../utils/constants'
+
+export function parseHeosMessageString(
+	messageString: string | undefined
+): HeosResponseMessage | string | void {
+	if (typeof messageString === 'string') {
+		const messageObject: HeosResponseMessage = {}
+		let keyAndValuePairs: string[] = messageString.split(
+			FIRST_LEVEL_HEOS_RESPONSE_MESSAGE_SPLITTER
+		)
+
+		// A single message (without message1&message2 pattern)
+		if (keyAndValuePairs.length === 1) {
+			keyAndValuePairs = [messageString]
+		}
+
+		keyAndValuePairs.forEach((keyAndValuePair, i) => {
+			const keyAndValue: string[] = keyAndValuePair.split(
+				SECOND_LEVEL_HEOS_RESPONSE_MESSAGE_SPLITTER
+			)
+
+			// Accept only key=value pattern
+			if (keyAndValue.length === 2) {
+				const value = keyAndValue[1]
+				messageObject[keyAndValue[0]] = !isNaN(Number(value)) ? Number(value) : value
+			}
+		})
+
+		// If the message doesn't have any splitter return as is
```

Comment:
I think that the message always has key-value pairs. Maybe this edge case is not necessary. What do you think?

Reply:
Actually there is...
`heos://system/sign_out`
```
{
 "heos": {
 "command": "system/sign_out ",
 "result": "success",
 "message": "signed_out"
 }
}
```
But there is also 
` heos://system/sign_in?un=heos_username&pw=heos_password`
```
{
 "heos": {
 "command": "system/sign_in ",
 "result": "success",
 "message": "signed_in&un=<current user name>"
 }
}
```
Messy

Reply:
What about this
```ts
export type HeosResponseMessage = {
	original: string,
	[key: string]: string | number 
}
```

Reply:
Very good point! I had not understood that message could be just a string. This complicates everything of course.

So to summarize we have three cases:
1. Message is only an unformatted string.
2. Message is a string with key-value pair.
3. There is no message.

The question is what kind of output we want in every case. I see it as two options.
1. We have a single generic type that encapsulates every input case.
2. There is one output type for every input type, totalling to 3 types.

Reply:
Of course, there are other options where there exists 2 output types.

Reply:
Also we have the case of the input being an empty string. So there are four input cases

Reply:
I think it would be possible to see `null`, `undefined`, `''` (empty string) all as being the same as a key-value formatted string, representing an "empty object".

This would reduce the input types to only two. The formatted and unformatted string.

Reply:
I personally think that would lead to a very nice API. What do you think?

Reply:
```ts
export type HeosResponseMessage = {
    unparsed: string,
    parsed: {
        [key: string]: string | number
    }
}
```

Should we also translate on/of to `true`/`false`?

Reply:
I really like that type! I assume that parsed is optional. Like this:

```ts
export type HeosResponseMessage = {
    unparsed: string,
    parsed?: {
        [key: string]: string | number
    }
}
```

This could cover every output case, which is nice. We could have only one output type for the function.

It is a good point about translating `on`/`off` to `true`/`false`. I do not know. Maybe begin with not translating any fields. I think it is a good idea to discuss this in a separate PR.

Reply:
I think both unparsed and parsed should be optional as the message might be missing completely.

About on/off -> true/false I agree it's better to discuss about it later on.

Reply:
Yeah, that is a good point, about having both optional. About on/off I think there might be more cases that would be nice to include. It could benefit from being its own discussion.

About the type, instead of using optionals I think it would be good to use `|` as otherwise uncorrect types would be possible. Like an object with `parsed` but not `unparsed`.

Like this:

```ts
export type HeosResponseMessage = {
    unparsed: string,
    parsed?: {
        [key: string]: string | number
    }
} | {
    unparsed: string,
} | {}
```

I am conflicted about if the last type should be `{}`, `null` or `undefined`. I am also leaning towards there being no difference between an empty message or a nonexistent message. So that both would produce the second type.